applications/rag/frontend/container/application/rag_langchain/rag_chain.py

```diff
@@ -0,0 +1,119 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import logging
+
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+from langchain_core.runnables import RunnableParallel, RunnableLambda
+from langchain_core.runnables.history import RunnableWithMessageHistory
+
+from langchain_community.embeddings.huggingface import HuggingFaceEmbeddings
+from langchain_google_cloud_sql_pg import PostgresChatMessageHistory
+
+
+from application.utils import (
+    create_sync_postgres_engine
+)
+from application.rag_langchain.huggingface_inference_model import (
+    HuggingFaceCustomChatModel,
+)
+from application.vector_storages import CloudSQLVectorStore
+
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+
+CHAT_HISTORY_TABLE_NAME = os.environ.get("CHAT_HISTORY_TABLE_NAME", "message_store")
+
+QUESTION = "input"
+HISTORY = "chat_history"
+CONTEXT = "context"
+
+SENTENCE_TRANSFORMER_MODEL = "intfloat/multilingual-e5-small"  # Transformer to use for converting text chunks to vector embeddings
+
+template_str = """Answer the question given by the user in no more than 2 sentences. 
+Use the provided context to improve upon your previous answers. Stick to the facts and be brief. Avoid conversational format.
+\n\n
+Context: {context} 
+"""
+
+prompt = ChatPromptTemplate.from_messages(
+    [
+        ("system", template_str),
+        MessagesPlaceholder("chat_history"),
+        ("human", "{input}"),
+    ]
+)
+
+engine = create_sync_postgres_engine()
+
+
+def get_chat_history(session_id: str) -> PostgresChatMessageHistory:
+    history = PostgresChatMessageHistory.create_sync(
+        engine, session_id=session_id, table_name=CHAT_HISTORY_TABLE_NAME
+    )
+
+    logging.info(
+        f"Retrieving history for session {session_id} with {len(history.messages)}"
+    )
+    return history
+
+
+def clear_chat_history(session_id: str):
+    history = PostgresChatMessageHistory.create_sync(
+        engine, session_id=session_id, table_name=CHAT_HISTORY_TABLE_NAME
+    )
+    history.clear()
+
+
+def create_chain() -> RunnableWithMessageHistory:
+    try:
+        model = HuggingFaceCustomChatModel()
+
+        langchain_embed = HuggingFaceEmbeddings(model_name=SENTENCE_TRANSFORMER_MODEL)
+        vector_store = CloudSQLVectorStore(langchain_embed, engine)
+
+        retriever = vector_store.as_retriever()
+
+        setup_and_retrieval = RunnableParallel(
+            {
+                "context": retriever,
+                QUESTION: RunnableLambda(lambda d: d[QUESTION]),
+                HISTORY: RunnableLambda(lambda d: d[HISTORY]),
+            }
+        )
+
+        chain = setup_and_retrieval | prompt | model
+        chain_with_history = RunnableWithMessageHistory(

```

Comment:
Can you add some comments around how the `chain.invoke` works with this `chain_with_history` and user input? Especially around how `setup_and_retrieval` component works.